nestkernel/recording_backend_mpi.h

```diff
@@ -0,0 +1,156 @@
+3) Receive at each ending of the run (true, 1, CXX_BOOL, 0, 0)
+4) Send shape of the data of the run (shape, 1,I NT, 0, 0)
+5) Send data of the data of the run (data, shape, DOUBLE, 0, 0)
+6) Send at each ending of the run (true, 1, CXX_BOOL, 0, 1)
+7) Send at this en of the simulation (true, 1, CXX_BOOL, 0, 2)
+
+
+Data format
++++++++++++
+
+The data which will be exchanged depends on the recording device.
+
+Parameter summary
++++++++++++++++++
+
+.. glossary::
+
+ port_file
+   Shared file with ports with the format path+label+id+.txt
+   Where path = kernel().io_manager.get_data_path() and the
+   label and id are the ones provided for the device.
+
+EndUserDocs */
+
+namespace nest
+{
+
+/**
+ * A recording backend for sending information with MPI.
+ */
+class RecordingBackendMPI : public RecordingBackend
+{
+public:
+  RecordingBackendMPI();
+  ~RecordingBackendMPI() throw();
+
+  void initialize() override;
+  void finalize() override;
+
+  void enroll( const RecordingDevice& device, const DictionaryDatum& params ) override;
+
+  void disenroll( const RecordingDevice& device ) override;
+
+  void set_value_names( const RecordingDevice& device,
+    const std::vector< Name >& double_value_names,
+    const std::vector< Name >& long_value_names ) override;
+
+  void cleanup() override;
+
+  void prepare() override;
+
+  void write( const RecordingDevice&, const Event&, const std::vector< double >&, const std::vector< long >& ) override;
+
+  void set_status( const DictionaryDatum& ) override;
+
+  void get_status( DictionaryDatum& ) const override;
+
+  void pre_run_hook() override;
+
+  void post_run_hook() override;
+
+  void post_step_hook() override;
+
+  void check_device_status( const DictionaryDatum& ) const override;
+  void get_device_defaults( DictionaryDatum& ) const override;
+  void get_device_status( const RecordingDevice& device, DictionaryDatum& params_dictionary ) const override;
+
+private:
+  bool enrolled_;
+  bool prepared_;
+  /**
+   * Buffer for saving events before they are sent.
+   * The buffer has 3 dimensions : thread_id, MPI_communicator_index and number of events
+   * elements. The events elements are described as an array with three components: id of device, id of neurons and data
+   * ( one double )
+   */
+  std::vector< std::vector< std::vector< std::array< double, 3 > > > > buffer_;
+  /**
+   * A map for the enrolled devices. We have a vector with one map per local
+   * thread. The map associates the node ID of a device on a given thread
+   * with its MPI index and device. Only the master thread have the have a valid MPI communicator pointer.
+  */
+  typedef std::vector< std::map< index, std::tuple< int, MPI_Comm*, const RecordingDevice* > > > device_map;
+  device_map devices_;
+  /**
+   * A map of MPI communicators used by the master thread for the MPI communication.
+   * This map contains also the number devices linked to each MPI communicator.
```

Comment:
I think this is still a bit too vague. I suggest to explicitly list what the tuple contains:
```suggestion
   * The values of the map are tuples containing the index of the MPI communicator, the MPI communicator itself,
   * and the number of devices linked to that MPI communicator.
```
Please verify that I have correctly guessed the contents of the tuple. :)

Reply:
it's correct